src/meta_data.c

```diff
@@ -18,37 +22,148 @@
+    bdf = strndup(slash1 + 1, bdf_str_len);
+    bdf[bdf_str_len] = '\0';
+
+    colon1 = strchr(bdf, ':');
+    colon1 += 1;
+    colon2 = strchr(colon1, ':');
+    null = colon2;
+    colon2 += 1;
+    *null = '\0';
+    dot = strchr(colon2, '.');
+    null = dot;
+    dot += 1;
+    *null = '\0';
+
+    x = atoi(colon1 + 1);
+    y = atoi(colon2);
+    z = atoi(dot);
+
+    ret |= z;
+    ret |= y << 8;
+    ret |= x << 16;
+
+    free(bdf);
+    return ret;
+}
+
+static int get_mac_addr(uint8_t *mac)
 {
     DIR *d;
+    char *link_buf;
+    char link_path[1024];
+    size_t link_buf_size = PATH_MAX;
     struct dirent *dir;
 
+    struct bdf_t {
+        char        iface[32];
+        uint32_t    bdf_num;
+    };
+
+    struct bdf_t bdf = {
+        .bdf_num = 0UL,
+    };
+
     if (access(SYS_CLASS_NET_PATH, R_OK))
-        PRINT_ERR_RET("No network on host: %s\n", strerror(errno));
+        PRINT_ERR_RET("ERR: No network on sysfs: %s\n", strerror(errno));
+
+    link_buf = malloc(link_buf_size);
+    if (!link_buf)
+        PRINT_ERR_RET("ERR: Cannot allocate memory: %s\n", strerror(errno));
 
     if ((d = opendir(SYS_CLASS_NET_PATH)) == NULL)
-        PRINT_ERR_RET("Cannot open directory %s\n", SYS_CLASS_NET_PATH);
+        PRINT_ERR_RET("ERR: Cannot open directory %s\n", SYS_CLASS_NET_PATH);
 
     while ((dir = readdir(d)) != NULL) {
-        static int i;
-        printf("dir%d: %s\n", i++, dir->d_name);
+        uint32_t curr_bdf;
+        struct stat sb;
+
+        if (dir->d_name[0] == '.')
+            continue;
+
+        snprintf(link_path, 1024, "%s/%s", SYS_CLASS_NET_PATH, dir->d_name);
+        if (lstat(link_path, &sb) < 0) {
+            fprintf(stderr, "WARN: Cannot stat the path: %s\n", link_path);
+            break;
+        }
+
+        if (!S_ISLNK(sb.st_mode))
+            continue;
+
+        memset(link_buf, 0, link_buf_size);
+        readlink(link_path, link_buf, link_buf_size);
+
+        /* Exclude virtual network devices */
+        if (strstr(link_buf, "virtual"))
+            continue;
+
+        curr_bdf = get_bdf_from_path(link_buf);
+        if (bdf.bdf_num == 0UL || bdf.bdf_num > curr_bdf) {
+            bdf.bdf_num = curr_bdf;
+            snprintf(bdf.iface, 32, "%s", (strrchr(link_buf, '/') + 1));
+        } 
     }
 
+    free(link_buf);
     closedir(d);
 
-    return 0;
+    if (bdf.bdf_num != 0) {
+        obtain_mac_from_iface(bdf.iface, mac);
+        return 0;
+    } else {
+        fprintf(stderr, "WARN: Cannot find any physical network interface in sysfs\n");
+        return -1;
+    }
 }
 
-int get_meta_data(const struct meta_data *meta)
+int get_meta_data(struct meta_data *meta)
 {
+    meta->header_version = META_H_VERSION;
+
     if (get_mac_addr(meta->mac) < 0)
         return -1;
 
+    printf("MAC: %2X:%2X:%2X:%2X:%2X:%2X\n",
+            meta->mac[0], meta->mac[1], meta->mac[2], meta->mac[3], meta->mac[4], meta->mac[5]);
+
+    get_dmi_system_info(meta);
```

Comment:
Please check return value. 
```suggestion
    if (get_dmi_system_info(meta))
        return -1;
```

Reply:
Or just `return get_dmi_system_info(meta);`.

Reply:
Fixed.